### ISSUE TITLE
Why can the Triplet_Loss(X, Y) function calculate KL divergence between X and Y

### ISSUE THREAD

**Issue by BlackAngel-511** [kam1107/RealnessGAN#6]
When I am reading the code, I found that you utilize Triplet_Loss(feat_real, feat_fake) in  line 56 of learnG.py to implement the expression "E[KL(D(x)||D(G(z)))]" . The Triplet_Loss is  a object of class CategoricalLoss, so I read its forward function. But  I don't have the ability to understand how it works and the comment in code is few. I have not  way to further learn its principle. 

Could you explain why Triplet_Loss function can measure KL divergence between two distribute and purpose of third parameter "skewness" or add more comment in the class CategoricalLoss

Thank you very much！

**Comment by kam1107** [comment]
The KL divergence between two distributions is calculated by https://github.com/kam1107/RealnessGAN/blob/3a3b07faa5944edb219b9cb535a7d0de207f7019/training/loss.py#L51

'Skewness' is a parameter we use to control the divergence b/w two anchors. For example, when comparing D(x_real) with A1, we can further skew A1 to right by setting skewness to 1; similarly,  A0 can be further skewed to left by setting skewness to -1 when comparing with D(x_fake).

**Comment by krishnan-meep** [comment]
Couldn't the KL Divergences be calculated using (p * (p / q).log()).sum(), where p is either A1 or A0, and q is the discriminator output vector? If you removed the skewing is this what you would use?

**Comment by kam1107** [comment]
Yes. Skewness is a hyperparam in the experiments. It is used to control the difference b/w anchors. After the anchors being reshaped/skewed, the KL divergence is computed as usual.

**Comment by BlackAngel-511** [comment]
@kam1107  When skewness was set zero, I get a outcome through Triplet_Loss(X, Y) . But it is different to the outcome through the method torch.mean(torch.sum(X * X / Y).log())). What has caused this problem. Thank you!

**Comment by kam1107** [comment]
This is because we ignore the constant term in KL divergence as it does not contribute to gradient.
